### PR TITLE
Support for expansion-class modPC hardware, a modPC hardware item required for the atmos scanning app, and reworking ID card slots in a related fashion

### PR DESCRIPTION
**Pull request by zxaber**
## About The Pull Request
- Adds expansion bays to modular computers. These are a limited number of bays that can hold any expansion-type item. For example, a single bay could be used for the AI card slot, a printer, an APC-siphoning recharger, a sensor package required by specific programs, or another ID card slot. These are distinct from hardware such as processors or hard drives in that they should be considered "extra" hardware not required for most basic computer tasks. Tablets have 3 expansion bays, while laptops and consoles have four. PDAs, when they get moved over, are planned to have two (and one will be taken up by the cart reader). Due to how hardware is addressed, expansion bays still do not allow for duplicate hardware.

- All of the above examples have been moved to expansion bays, save for the card reader which has been split.

- Added a sensor package hardware item. The item itself does nothing, but is now required by AtmoZphere and possibly future programs. Atmos techs and command staff will receive tablets with this hardware preinstalled.

- The card reader has been split into two forms; The primary card reader and the secondary card reader. Each card reader can only hold one card (instead of two), and the secondary card reader requires an expansion bay. Primary card readers are standard issue in all modular computer devices, while secondary card readers are only pre-installed in devices that had a card reader prior. Card reader items can be swapped between the two bay types using a screwdriver.

- Programs can no longer use the user's worn ID for logins. IDs must be inserted into the primary ID slot.

- Supporting changes for the above across a number of files to ensure the rest of the game still works.

## Why It's Good For The Game
- Better control over hardware choices will help move modular computers into being actually modular. A user may decide, for example, that a sensor package is really important to them and they can do without a printer. Adding more hardware will aid in this being a meaningful choice, but for now, five items exist to play with the idea.

- The card reader change was made for a few reasons. Primarily, with the planned changes to PDAs and how I want to work them and the rest of the modular computers to fit together, I believe it will be helpful to (somewhat) guarantee a device can hold an ID. However, simply adding in the card slot hardware to every device essentially removes a hardware option as both a price sink and also a hardware choice. As such, I've split the function into two steps; For logging into your device (and using it to bump-open doors), the primary slot will work. If you require two cards (ID modification being the main example), you'll need a secondary ID slot and an open expansion bay. Alternatively, you can use the second slot to benefit from the combined access of two IDs (just like before).

- Removal of programs being able to just read your worn ID to get login info is something I've been wanting to do for a while. It's not super clear to the user why the tablet knows who you are, and picking up a device such as a tablet should leave it in the same state it was in when the prior user dropped it.

## Changelog
:cl:
add: Added support for modular PC expansion bays, or hardware bays designed to be used by nonessential hardware. Devices have a limited number of bays (tablets less than others).
balance: The AI card slot, printer, and APC-siphoning recharger are now expansion bay hardware.
balance: AtmoZphere, the modular PC atmos scanner, now requires a new expansion bay hardware item - the sensor package. Certain shift-start tablets have been adjusted to reflect this.
refactor: Modular PC card readers have been split -- The primary ID card reader, which comes pre-installed on nearly all devices, and a secondary ID card reader which is an expansion bay hardware and will need to be purchased separately. Certain shift-start devices have a secondary reader pre-installed.
fix: Buying a tablet from the Laptop/Tablet vendor, and paying extra for the nano printer, will now attach the correct printer rather than attempt to stuff a laptop printer into a tablet and silently fail but still take your cash.
/:cl:

### CODE REVIEW

File: code/modules/modular_computers/computers/item/computer.dm [hunk 1]
```diff
@@ -40,6 +40,8 @@
 	// Optional hardware (improves functionality, but is not critical for computer to work)
 
 	var/list/all_components = list()						// List of "connection ports" in this computer and the components with which they are plugged
+	var/list/expansion_bays = list()						// List of extra hardware slots that can be used modularly
```
Comment (Rohesie) [inline code/modules/modular_computers/computers/item/computer.dm:43]: Could you turn this into a lazy list?

File: code/modules/modular_computers/file_system/program.dm [hunk 2]
```diff
@@ -82,10 +82,18 @@
...
+  *ID must be inserted into a card slot to be read. If the program is not currently installed (as is the case when
+  *NT Software Hub is checking available software), a list can be given to be used instead.
+  *Arguments:
+  *user is a ref of the mob using the device.
+  *loud is a bool deciding if this proc should use to_chats
+  *access_to_check is an access level that will be checked against the ID
+  *transfer, if TRUE and access_to_check is null, will tell this proc to use the program's transfer_access in place of access_to_check
+  *access can contain a list of access numbers to check against. If access is not empty, it will be used istead of checking any inserted ID.
+*/
+/datum/computer_file/program/proc/can_run(mob/user, loud = FALSE, access_to_check, transfer = FALSE, var/list/access = list())
```
Comment (Rohesie) [inline code/modules/modular_computers/file_system/program.dm:96]:
```suggestion
/datum/computer_file/program/proc/can_run(mob/user, loud = FALSE, access_to_check, transfer = FALSE, var/list/access)
```

File: code/modules/modular_computers/file_system/program.dm [hunk 3]
```diff
@@ -104,29 +112,24 @@
 	if(issilicon(user))
 		return TRUE
 
-	if(ishuman(user))
+	if(!access.len)
```
Comment (Rohesie) [inline code/modules/modular_computers/file_system/program.dm:115]:
```suggestion
	if(!length(access))
```

File: code/modules/modular_computers/file_system/programs/atmosscan.dm [hunk 4]
```diff
@@ -7,11 +7,20 @@
 	size = 4
 	tgui_id = "NtosAtmos"
 
+/datum/computer_file/program/atmosscan/run_program(mob/living/user)
+	if (!..())
+		return FALSE
+	if(!computer?.get_modular_computer_part(MC_SENSORS)) //Giving a clue to users why the program is spitting out zeros.
+		to_chat(user, "<span class='warning'>\The [computer] flashes an error: \"hardware\\sensorpackage\\startup.bin -- file not found\".</span>")
+	return TRUE
```
Comment (Rohesie) [inline code/modules/modular_computers/file_system/programs/atmosscan.dm:15]:
```suggestion
/datum/computer_file/program/atmosscan/run_program(mob/living/user)
	. = ..()
	if (!.)
		return
	if(!computer?.get_modular_computer_part(MC_SENSORS)) //Giving a clue to users why the program is spitting out zeros.
		to_chat(user, "<span class='warning'>\The [computer] flashes an error: \"hardware\\sensorpackage\\startup.bin -- file not found\".</span>")
```

File: code/modules/modular_computers/hardware/card_slot.dm [hunk 5]
```diff
@@ -76,53 +63,58 @@
...
+	if(holder)
+		if(holder.active_program)
+			holder.active_program.event_idremoved(0)
+
+		for(var/I in holder.idle_threads)
+			var/datum/computer_file/program/P = I
+			P.event_idremoved(1)
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		H.sec_hud_set_ID()
```
Comment (Rohesie) [inline code/modules/modular_computers/hardware/card_slot.dm:86]:
```suggestion
		for(var/p in holder.idle_threads)
			var/datum/computer_file/program/computer_program = p
			computer_program.event_idremoved(1)
	if(ishuman(user))
		var/mob/living/carbon/human/human_user = user
		human_user.sec_hud_set_ID()
```
* Better var names.

File: code/modules/modular_computers/hardware/card_slot.dm [hunk 6]
```diff
@@ -76,53 +63,58 @@
 	return TRUE
 
 
-/obj/item/computer_hardware/card_slot/try_eject(slot=0, mob/living/user = null, forced = 0)
-	if(!stored_card && !stored_card2)
+/obj/item/computer_hardware/card_slot/try_eject(mob/living/user = null, forced = 0)
```
Comment (Rohesie) [inline code/modules/modular_computers/hardware/card_slot.dm:66]:
```suggestion
/obj/item/computer_hardware/card_slot/try_eject(mob/living/user = null, forced = FALSE)
```

File: code/modules/modular_computers/hardware/card_slot.dm [hunk 7]
```diff
@@ -76,53 +63,58 @@
...
 /obj/item/computer_hardware/card_slot/attackby(obj/item/I, mob/living/user)
 	if(..())
 		return
 	if(I.tool_behaviour == TOOL_SCREWDRIVER)
-		to_chat(user, "<span class='notice'>You press down on the manual eject button with \the [I].</span>")
-		try_eject(0,user)
-		return
+		if(stored_card)
+			to_chat(user, "<span class='notice'>You press down on the manual eject button with \the [I].</span>")
+			try_eject(0,user)
```
Comment (Rohesie) [inline code/modules/modular_computers/hardware/card_slot.dm:97]:
```suggestion
			try_eject(user)
```
Better look for all calls of `try_eject` and make sure you remove the first argument if it still refers to `slot`

Comment (zxaber) [reply code/modules/modular_computers/hardware/card_slot.dm]: Oof, I went and missed the one that was in the same file.

Did a search, and removed any reference to the slot number in the rest of the code. I also removed slot references in the `try_eject()` proc for the AI reader and battery module (both of which did not use slots but I assume had the arg to be consistent with the card reader).

File: code/modules/modular_computers/computers/item/computer_components.dm [hunk 8]
```diff
@@ -7,10 +7,10 @@
 		return FALSE
 
 	if(H.expansion_hw)
-		if(expansion_bays.len >= max_bays)
+		if(length(expansion_bays) >= max_bays)
```
Comment (Rohesie) [inline code/modules/modular_computers/computers/item/computer_components.dm:10]:
```suggestion
		if(LAZYLEN(expansion_bays) >= max_bays)
```
This is the same as `length()` but it makes it clear this is a lazylist

File: code/modules/modular_computers/computers/item/computer.dm [hunk 9]
```diff
@@ -40,6 +40,8 @@
 	// Optional hardware (improves functionality, but is not critical for computer to work)
 
 	var/list/all_components = list()						// List of "connection ports" in this computer and the components with which they are plugged
+	var/list/expansion_bays									// Lazy List of extra hardware slots that can be used modularly
+	var/max_bays = 0										// Number of total expansion bays this computer has available.
```
Comment (Rohesie) [inline code/modules/modular_computers/computers/item/computer.dm:44]:
```suggestion
	/// Lazy List of extra hardware slots that can be used modularly.
	var/list/expansion_bays
	/// Number of total expansion bays this computer has available.
	var/max_bays = 0
```

**Comment by Rohesie** [comment]
Fix the bundle conflict, test locally, and if working as intended it's ready to go in.

**Pull request merged**